### ISSUE TITLE
oversampling for unbalanced data ?

### ISSUE THREAD

**Issue by Sandy4321** [nlarusstone/corels#27]
What can be done for unbalanced data?
for example :
number of target "yes" is 200
but number of target "no" is 500000

Oversampling , meaning replicating records with target "yes" can be the solution

When oversampling used only one time, it helps little bit
So it will be
number of target yes is 400
number of target no is 500000

But second replicating surprisingly is not helping  at all, relatively to first replicating
second replicating confusion matrix is the same as  first replicating confusion matrix
so when it is
number rows for  target yes is 600
number rows  for  target no is 500000

then performance is the same as when
number of target yes is 400
number of target no is 500000

The questions are:
1
do you remove identical rows?
2
Do you have weighting for particular rows?
for example
rows with targets yes is may have more influence than rows targets "no"
Then rows weighting can be used for unbalanced data?

Thanks

**Comment by cynrudin** [comment]
Hi - if you want to achieve approximately equal accuracy on the two targets (usually we call them “positive” and “negative”), you need to replicate the “positive” ’s so that you have approximately the same number of “positives” and “negatives”.

**Comment by Sandy4321** [comment]
Thanks for soon answer
But it is exactly what I'm doing 
I gradually increase number of minorities
In your terminology number of positives
So I expect to see difference when I doubled number of positives
But do not see difference
Does your answer mean, that I should increase so significantly that number of positives and negatives will be the same
And that doubling is not enough to make difference?
Also that effect from increasing positives only visible in non linear manner?
Meaning when number of positives increased so significantly that number of positives and negatives became the same , only then effect from positives increase addruptively become noticable?
So not gradually but addruptively? 
In other words, algorithm is not sensible to small number of positives
So it is the same if we have 100 positives or 200
When number of negatives is 300000?

**Comment by cynrudin** [comment]
Hi, You will not see a difference until you add a lot more positives with the level of imbalance you currently have. The loss function in CORELS weights each point equally. So, until you add a lot more positives, the optimal classifier would prefer just to sacrifice all of the positives to achieve a low classification error. As you know, CORELS is programmed to be sparse. If the regularization parameter lambda = 0.01 and less than 1% of your observations are positives, then CORELS will never even create a positive leaf. The algorithm will terminate immediately and the solution will be optimal according to CORELS’ objective. If you care about getting some level of accuracy on your positives, you will have to make each positive much more important than each negative by duplicating the positives. Cheers, Cynthia

**Comment by Sandy4321** [comment]
Great 
Thanks for soon answer
I see
Hopefully in future you will add something like F1 metric or something else
To treat unbalanced data natively
Or as usual you suprising us from  one breakthrough to another breakthrough
Then my intuition tells something big coming?
Something even better than Corels?

**Comment by cynrudin** [comment]
Thanks - we are trying to improve the OSDT algorithm so that it handles more general loss functions. OSDT doesn’t scale nearly as well as CORELS, but it does allow full binary-split decision trees. https://github.com/xiyanghu/OSDT We are working on CORELS improvements in parallel, but not as heavily at the moment so that we can get our ODST stuff off the ground.

**Comment by Sandy4321** [comment]
1
great news about OSDT, thanks
only as usual I need to ask : if this OSDT code https://github.com/xiyanghu/OSDT is suitable to Windows computer?
 hopefully answer is yes, especially since majority in now days using Windows
2
What is about very similar features , kind of multicolinearity for corels and OSDT
should be similar features be joint to one feature or one of them should be removed
and what function from
 https://docs.scipy.org/doc/scipy/reference/spatial.distance.html
is the best to identify these similar features
per
Distance functions between two boolean vectors (representing sets) u and v. As in the case of numerical vectors, pdist is more efficient for computing the distances between all pairs.

dice(u, v[, w])

Compute the Dice dissimilarity between two boolean 1-D arrays.

hamming(u, v[, w])

Compute the Hamming distance between two 1-D arrays.

jaccard(u, v[, w])

Compute the Jaccard-Needham dissimilarity between two boolean 1-D arrays.

kulsinski(u, v[, w])

Compute the Kulsinski dissimilarity between two boolean 1-D arrays.

rogerstanimoto(u, v[, w])

Compute the Rogers-Tanimoto dissimilarity between two boolean 1-D arrays.

russellrao(u, v[, w])

Compute the Russell-Rao dissimilarity between two boolean 1-D arrays.

sokalmichener(u, v[, w])

Compute the Sokal-Michener dissimilarity between two boolean 1-D arrays.

sokalsneath(u, v[, w])

Compute the Sokal-Sneath dissimilarity between two boolean 1-D arrays.

yule(u, v[, w])

Compute the Yule dissimilarity between two boolean 1-D arrays.



Thank you very much in advance for taking care

**Comment by cynrudin** [comment]
Hi - OSDT is not as advanced as CORELS. It is what we are working on lately, and it solves a harder problem than CORELS so it doesn’t scale the way that CORELS does. I do not believe OSDT is set up for windows, I don’t think Sean was concerned about that...
Indeed multicollinearity is going to cause an issue for any of these methods (although we are working on improving exactly this issue… sigh…) I think either correlation or Hamming distance should be fine for removing redundant features. I’m not familiar with most of the distance metrics in that list!
Cheers,
Cynthia

**Comment by Sandy4321** [comment]
I don’t think Sean was concerned about that
Hello Sean
Can you please  make your project to be suitable for Windows.
We are vast majority who use Windows
So you significantly increase popularly of your code
And help many people
Corels also was in very beginning for Linux only
But it was adapted to Windows as well
So you can do the same
Thank you very much in advance

**Comment by xiyanghu** [comment]
Hi - It is totally OK to run OSDT on Windows. You need to install the `gmpy2` package by `pip install gmpy2`. I just tested the code on Windows, and it works well.
Cheers,
Sean

**Comment by cynrudin** [comment]
Thanks Sean. :)

**Comment by Sandy4321** [comment]
Hello Sean
I try as you recommend to install
gmp2 by 
pip install gmpy2

but get this 
Microsoft Windows [Version 10.0.17763.973]
(c) 2018 Microsoft Corporation. All rights reserved.

C:\WINDOWS\system32>pip install gmpy2
Collecting gmpy2
  Downloading https://files.pythonhosted.org/packages/90/f4/9a2e384b325b69bc5827b9a6510a8fb4a51698c915c06a3f25a86458892a/gmpy2-2.0.8.zip (280kB)
    100% |████████████████████████████████| 286kB 4.7MB/s
Building wheels for collected packages: gmpy2
  Building wheel for gmpy2 (setup.py) ... error
  Complete output from command C:\Users\User\Anaconda3\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\User\\AppData\\Local\\Temp\\pip-install-gg1ym961\\gmpy2\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d C:\Users\User\AppData\Local\Temp\pip-wheel-2rcyzblh --python-tag cp37:
  running bdist_wheel
  running build
  running build_ext
  building 'gmpy2' extension
  creating build
  creating build\temp.win-amd64-3.7
  creating build\temp.win-amd64-3.7\Release
  creating build\temp.win-amd64-3.7\Release\src
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MT -DMPIR -DWITHMPFR -DWITHMPC -IC:\Users\User\Anaconda3\include -IC:\Users\User\Anaconda3\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" /Tcsrc\gmpy2.c /Fobuild\temp.win-amd64-3.7\Release\src\gmpy2.obj
  gmpy2.c
  C:\Users\User\AppData\Local\Temp\pip-install-gg1ym961\gmpy2\src\gmpy.h(104): fatal error C1083: Cannot open include file: 'mpir.h': No such file or directory
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.22.27905\\bin\\HostX86\\x64\\cl.exe' failed with exit status 2

  ----------------------------------------
  Failed building wheel for gmpy2
  Running setup.py clean for gmpy2
Failed to build gmpy2
Installing collected packages: gmpy2
  Running setup.py install for gmpy2 ... error
    Complete output from command C:\Users\User\Anaconda3\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\User\\AppData\\Local\\Temp\\pip-install-gg1ym961\\gmpy2\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\User\AppData\Local\Temp\pip-record-ogt62v22\install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_ext
    building 'gmpy2' extension
    creating build
    creating build\temp.win-amd64-3.7
    creating build\temp.win-amd64-3.7\Release
    creating build\temp.win-amd64-3.7\Release\src
    C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MT -DMPIR -DWITHMPFR -DWITHMPC -IC:\Users\User\Anaconda3\include -IC:\Users\User\Anaconda3\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.22.27905\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" /Tcsrc\gmpy2.c /Fobuild\temp.win-amd64-3.7\Release\src\gmpy2.obj
    gmpy2.c
    C:\Users\User\AppData\Local\Temp\pip-install-gg1ym961\gmpy2\src\gmpy.h(104): fatal error C1083: Cannot open include file: 'mpir.h': No such file or directory
    error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.22.27905\\bin\\HostX86\\x64\\cl.exe' failed with exit status 2

    ----------------------------------------
Command "C:\Users\User\Anaconda3\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\User\\AppData\\Local\\Temp\\pip-install-gg1ym961\\gmpy2\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\User\AppData\Local\Temp\pip-record-ogt62v22\install-record.txt --single-version-externally-managed --compile" failed with error code 1 in C:\Users\User\AppData\Local\Temp\pip-install-gg1ym961\gmpy2\

C:\WINDOWS\system32>

**Comment by Sandy4321** [comment]
same story for 
pip install gmpy2==2.1.0b3
per
https://pypi.org/project/gmpy2/2.1.0b3/

**Comment by Sandy4321** [comment]
Sean
and by the way 
test file has call for monk2-train.csv
but this file in not in folder data

**Comment by Sandy4321** [comment]
Sean
this can be useful
https://stackoverflow.com/questions/40075271/gmpy2-not-installing-mpir-h-not-found
https://www.lfd.uci.edu/~gohlke/pythonlibs/
GMPY supports fast multiple-precision arithmetic.
Wraps the MPIR, MPFR, and MPC libraries.
gmpy‑1.17‑cp34‑none‑win_amd64.whl
gmpy‑1.17‑cp34‑none‑win32.whl
gmpy‑1.17‑cp27‑none‑win_amd64.whl
gmpy‑1.17‑cp27‑none‑win32.whl
gmpy2‑2.0.8‑cp38‑cp38‑win_amd64.whl
gmpy2‑2.0.8‑cp38‑cp38‑win32.whl
gmpy2‑2.0.8‑cp37‑cp37m‑win_amd64.whl

**Comment by xiyanghu** [comment]
Hi - I used “conda install gmpy2” to install the gmpy2 package on my PC. If you have already installed Anaconda, you can try this command. It worked well on my side.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by Sandy4321** [comment]
great thanks
also this line 
in test file will couse errors
test_accuracy('../data/preprocessed/compas-binary.csv', lambs=[0.025, 0.01, 0.005, 0.001, 0.0005],
              file_CART=r'./accuracy/cart_compas.txt', file_OSDT=r'./accuracy/osdt_compas.txt')

is it possible to add folder accuracy and files like cart_compas.txt?

**Comment by xiyanghu** [comment]
Just added this dataset.

________________________________

**Comment by Sandy4321** [comment]
OK
use winpython 
so installed by
gmpy2-2.0.8-cp37-cp37m-win_amd64.whl

now can you share simple test file
to check if all works please
something like in
https://github.com/fingoldin/pycorels
from corels import CorelsClassifier

# ["loud", "samples"] is the most verbose setting possible
C = CorelsClassifier(max_card=2, c=0.0, verbosity=["loud", "samples"])

# 4 samples, 3 features
X = [[1, 0, 1], [0, 0, 0], [1, 1, 0], [0, 1, 0]]
y = [1, 0, 0, 1]
# Feature names
features = ["Mac User", "Likes Pie", "Age < 20"]

# Fit the model
C.fit(X, y, features=features, prediction_name="Has a dirty computer")

# Print the resulting rulelist
print(C.rl())

# Predict on the training set
print(C.predict(X))

**Comment by Sandy4321** [comment]
now error is
OSDT-master\src\test_accuracy_1643.py", line 44, in test_accuracy
  with open(file_CART, 'a+') as f:

builtins.FileNotFoundError: [Errno 2] No such file or directory: './accuracy/cart_compas.txt'

**Comment by Sandy4321** [comment]
also after commenting file read lines

    #with open(file_CART, 'a+') as f:
        #f.write(";".join(["fold", "lamb", "nleaves", "trainaccu_CART", "testaccu_CART"]) + '\n')
    #with open(file_OSDT, 'a+') as f:
        #f.write(";".join(["fold", "lamb", "nleaves", "trainaccu_OSDT", "testaccu_OSDT", "totaltime", "time_c", "leaves_c"]) + '\n 


error is ( this file name looks stange : compas-binary.csv.train1.csv)
OSDT-master\src\test_accuracy_1643.py", line 54, in test_accuracy
  data_train = pd.DataFrame(pd.read_csv(file_train, sep=";"))
File "C:\Users\User\Anaconda3\Lib\site-packages\pandas\io\parsers.py", line 702, in parser_f
  return _read(filepath_or_buffer, kwds)
File "C:\Users\User\Anaconda3\Lib\site-packages\pandas\io\parsers.py", line 429, in _read
  parser = TextFileReader(filepath_or_buffer, **kwds)
File "C:\Users\User\Anaconda3\Lib\site-packages\pandas\io\parsers.py", line 895, in __init__
  self._make_engine(self.engine)
File "C:\Users\User\Anaconda3\Lib\site-packages\pandas\io\parsers.py", line 1122, in _make_engine
  self._engine = CParserWrapper(self.f, **self.options)
File "C:\Users\User\Anaconda3\Lib\site-packages\pandas\io\parsers.py", line 1853, in __init__
  self._reader = parsers.TextReader(src, **kwds)
File "C:\Users\User\Anaconda3\Lib\site-packages\pandas\_libs\parsers.cp37-win_amd64.pyd", line 387, in pandas._libs.parsers.TextReader.__cinit__
File "C:\Users\User\Anaconda3\Lib\site-packages\pandas\_libs\parsers.cp37-win_amd64.pyd", line 705, in pandas._libs.parsers.TextReader._setup_parser_source

builtins.FileNotFoundError: [Errno 2] File b'../data/preprocessed/compas-binary.csv.train1.csv' does not exist: b'../data/preprocessed/compas-binary.csv.train1.csv'

**Comment by xiyanghu** [comment]
I just pushed some updates to the OSDT repository. You can clone it and try again. It should work well now.

**Comment by Sandy4321** [comment]
Sean, 
great thanks
it started run without error
but it takes a lot of time
can please share some simple but meaningful toy example test file  like
to make sure it works
for example like 
https://github.com/fingoldin/pycorels
Toy dataset (See picture example above)
from corels import CorelsClassifier

# ["loud", "samples"] is the most verbose setting possible
C = CorelsClassifier(max_card=2, c=0.0, verbosity=["loud", "samples"])

# 4 samples, 3 features
X = [[1, 0, 1], [0, 0, 0], [1, 1, 0], [0, 1, 0]]
y = [1, 0, 0, 1]
# Feature names
features = ["Mac User", "Likes Pie", "Age < 20"]

# Fit the model
C.fit(X, y, features=features, prediction_name="Has a dirty computer")

# Print the resulting rulelist
print(C.rl())

# Predict on the training set
print(C.predict(X))

**Comment by nlarusstone** [comment]
Closing -- discussion can continue in the OSDT repo
